internal/controller/etcdcluster_controller.go

```diff
@@ -47,17 +59,193 @@ type EtcdClusterReconciler struct {
+	}
+
+	memberCnt := 0
+	if memberListResp != nil {
+		memberCnt = len(memberListResp.Members)
+	}
+
+	/* if memberCnt != len(healthInfos) {
+		// Only proceed when all members are healthy
+		return ctrl.Result{}, fmt.Errorf("memberCnt (%d) isn't equal to healthy member count (%d)", memberCnt, len(healthInfos))
+	} */
 
-	// TODO(user): your logic here
+	var learnerStatus *clientv3.StatusResponse
+	var learner uint64
+	var leaderStatus *clientv3.StatusResponse
+	if memberCnt > 0 {
+		// Find the leader status
+		_, leaderStatus = etcdutils.FindLeaderStatus(healthInfos, logger)
 
+		if leaderStatus == nil {
+			// If the leader is not available, let's wait for the leader to be elected
+			return ctrl.Result{}, fmt.Errorf("couldn't find leader, memberCnt: %d", memberCnt)
+		}
+
+		learner, learnerStatus = etcdutils.FindLearnerStatus(healthInfos, logger)
+
+		if learner > 0 {
+			// There is at least one learner. Let's try to promote it or wait
+			// Find the learner status
+			logger.Info("Learner found", "learnedID", learner)
+			logger.Info("Learner status", "learnerStatus", learnerStatus)
+			if etcdutils.IsLearnerReady(leaderStatus, learnerStatus) {
+				logger.Info("Learner is ready to be promoted to voting member", "learnerID", learner)
+				logger.Info("Promoting the learner member", "learnerID", learner)
+				eps := clientEndpointsFromStatefulsets(sts)
+				eps = eps[:(len(eps) - 1)]
+				err = etcdutils.PromoteLearner(eps, learner)
+				if err != nil {
+					// The member is not promoted yet, so we error out
+					return ctrl.Result{}, err
+				}
+			} else {
+				// Learner is not yet ready. We can't add another learner or proceed further until this one is promoted
+				// So let's requeue
+				logger.Info("The learner member isn't ready to be promoted yet", "learnerID", learner)
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+		}
+	}
+
+	//  Check if the size of the stateful set is less than expected size
+	//  Or if there is a pending learner to be promoted
+	if *sts.Spec.Replicas < int32(etcdCluster.Spec.Size) && memberCnt > 0 {
+		// If there is no more learner, then we can proceed to scale the cluster further.
+		// If there is no more member to add, the control will not reach here after the requeue
+
+		targetReplica := int(*sts.Spec.Replicas) + 1
+		_, peerURL := peerEndpointForOrdinalIndex(etcdCluster, int(*sts.Spec.Replicas)) // The index starts at 0, so we do not need to add 1
+		eps := clientEndpointsFromStatefulsets(sts)
+		logger.Info("[Scale out] adding a new learner member to etcd cluster", "peerURLs", peerURL)
+		if _, err := etcdutils.AddMember(eps, []string{peerURL}, true); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		logger.Info("Learner member added successfully", "peerURLs", peerURL)
+
+		sts, err = createOrPatchStatefulSet(ctx, logger, etcdCluster, r.Client, int32(targetReplica), r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Minor enhancement: We only need to call `createOrPatchStatefulSet` once, currently you call it twice in two if branches.

The rough psedo code is something like below,

```
if replica == expectedSize {
    return
}

if replica < expectedSize {
    call AddMember
} else{
    call memberRemove
}

call createOrPatchStatefulSet with the targetReplica
```